text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+
+This RFC introduces 4 new traits.
+
+The core of this RFC is the `Store` trait:
+
+```rust
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store {
+    /// Handle to a block of memory.
+    type Handle: Copy;
+
+    /// Returns a dangling handle, always invalid.
+    fn dangling() -> Self::Handle;
+
+    /// Return a pointer to the block of memory associated to `handle`.
+    unsafe fn resolve(&self, handle: Self::Handle) -> NonNull<u8>;
+
+    //  The following methods are similar to `Allocator`, reformulated in terms of `Self::Handle`.
+
+    /// Allocates a new handle to a block of memory. On success, invalidates any existing handle.
+    fn allocate(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Deallocates a handle.
+    unsafe fn deallocate(&self, handle: Self::Handle, layout: Layout);
+
+    /// Grows the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn grow(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Shrinks the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn shrink(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Allocates a new handle to a block of zeroed memory. On success, invalidates any existing handle.
+    fn allocate_zeroed(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+
+    /// Grows the block of memory associated to a handle with zeroed memory. On success, the handle is invalidated.
+    unsafe fn grow_zeroed(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+}
+```
+
+_Note:  full-featured documentation of the trait and methods can be found in the companion repository at_
+        https://github.com/matthieu-m/store/blob/main/src/interface.rs.
+
+The `Store` trait is supplemented by 3 additional marker traits, providing extra guarantees:
+
+```rust
+/// A refinement of `Store` which does not invalidate existing handles on allocation, and does not invalidate
+/// unrelated existing handles on deallocation.
+pub unsafe trait MultipleStore: Store {}
+
+/// A refinement of `Store` which does not invalidate existing pointers on allocation, resolution, or deallocation, but
+/// may invalidate them on moves.
+pub unsafe trait StableStore: Store {}
+
+/// A refinement of `Store` which does not invalidate existing pointers, not even on moves. That is, this refinement
+/// guarantees that the blocks of memory are pinned in memory.
+pub unsafe trait PinningStore: StableStore {}
```

Comment:
I am not sure I see the problem.

Just because `PinningStore` provides a guarantee does not necessarily mean that `Box<..., PinningStore>` does, since `Box` is free to impose extra constraints, and unicity is just one such constraint.

I would expect `Box` to have a `UniqueHandle<S::Handle, T>` which could offer the guarantees that `Unique<T>` currently offers.

Reply:
How would it actually do so, though? It's not just an API guarantee, but a property of the borrowing model that moving `Box` retags its pointer. There's no pointer in `(S::Handle, S)` *to* retag.

Reply:
I was more thinking that `Box` can keep the API guarantee _regardless_ of whether under the hood a retag occurs.

This may mean that certain checks or optimizations can only apply when the handle is an actual pointer, which is a bit sad, but something I'd be happy to live with.

Reply:
But again, I'm not talking about the API guarantee, I'm talking about opsem requirements. The retag is exactly what permits the optimizations, by making reuse of derived pointers invalid and UB, rather than just an API soundness violation.

When talking about compiler optimizations, it's not sufficient to say "the API prevents this" to introduce UB; you require "the language semantics prevent this."

Reply:
I'm sorry, I feel like I am missing something. I agree with the statements in general, I just don't see how they apply to the case at hand... possibly because I don't really _see_ a case at hand in the first place.

I think it would be easier for me to understand what you're trying to say, if we had a concrete case to discuss.

Could you please create a small snippet of code with `Box`, then:

- Explain at which point this "retag" happens with the current `Box`, and what UB it prevents or what optimizations it enables.
- Explain why no such "retag" would occur with a `StoreBox`.

As it stands... I'm just lost.

Reply:
I'll attempt to set up an example this weekend then. Tbf, it's quite a subtle property/issue, and it's relying on the (non) availability of compiler magic, so "more compiler magic" (and/or specialization, which, being unstable, is currently essentially just compiler magic by a more specific name) is always theoretically a possible solution. But we'd also like to continue to move towards `Box` being less special, not more. It would theoretically be possible to get `Box`-like retagging properties from building it on `&'static mut T` instead of `ptr::Unique<T>`. (And note that `ptr::Unique<T>` is just an API wrapper around `ptr::NonNull<T>` presently; the magic semantics are on `Box<T>` instead.)

Reply:
building `Box` on `&'static mut T` won't work since `T` could have non-static lifetimes in it. e.g.
```rust
pub fn f<'a>(v: &'a str) -> Box<&'a str> {
    Box::new(v)
}
```

Reply:
Yes, building `Box` from `&'static mut` wouldn't be sufficient, since the variance would be wrong. The underlying point, though, is that the retag behavior of `Box<T>` (or `ptr::Unique<T>`, if the magic is moved to there) can be understood correctly as being analogous to the retag behavior of `&'static mut`[^1].

[^1]: With some small caveats around deallocation provenance; namely, protectors on `&mut` require it not to be deallocated within the scope of a function it is an argument to.

[Here's a worked example of the retag on `Box`](https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=4ae8dc2478212a5f778b323c3b0f1256). Read [the Tree Borrows explainer](https://perso.crans.org/vanille/treebor/) if you haven't yet. When `Box` is passed as a function argument, it gets a unique retag, invalidating derived pointer provenance, and justifying the application of LLVM `noalias`. The canonical forbidden case is `fn(Box<Cell<T>>, &Cell<T>)` where the box and reference alias, as this would permit writes through the reference to alias the box pointer.

To reiterate, this is not and never has been about purely safe usage of the `Box` interface, nor even purely _sound_ usage either; it's about what and when language UB is thrown, and thus the assumptions the optimizing compiler can make about the code. Being able to assume that `Box<T>` is a unique pointer which does not alias other pointers is a huge optimization potential and a primary justification of Rust's ownership system.

You could of course recover the UB condition by making a lang-magic `UniqueHandle<T, S>` analogous to `ptr::Unique<T>`. The reason I'm not enthused about that solution isn't so much that it's not decreasing the amount of `Box` magic, but that it's _increasing_ the amount of `Box` magic, because while `ptr::Unique<T>` is just a covariant pointer with write-active retagging semantics, `UniqueHandle<T, S>` needs to "do a specialization" and add retagging semantics to `AllocHandle` handles but leave other handles[^2] untouched.

[^2]: Notably, it can't do it to other pointer-containing handles, because it's perfectly allowable to have a pointer-carrying handle without a no-op `resolve`.

Reply:
Thank you for the explanation, I think I am starting to understand the core issue.

My understanding so far is that `Unique<T>` being a lang-item is what allows the compiler to _know_ about the ownership situation, and therefore apply `noalias` on moves of the container -- be it `Box` or `Vec` or `VecDeque`. Miri builds upon this knowledge by retagging provenance on moves, and thus detect UB violations.

With that in mind, I am not sure that _more_ magic is required, at least, no more than is already used with `std`. By using specialization, it should be possible to specialize the fields of `UniqueHandle<T, S>` for the case of `S: Storage<Handle = AllocHandle>` to `(Unique<T>, S)`.

_Note: I don't think that Rust supports specializing _types_, only _impls_. This would only mean two distinct types implementing a `UniqueHandle` trait and a `UniqueStore` trait specialized for `AllocHandle` with the appropriate `UniqueHandle` associated type. Bit heavier, but does not require any (more) compiler magic for now._

Reply:
It's actually `Box` which has the magic semantics attached currently rather than `Unique`, but effectively yes.

The weird specialization is pretty much exactly what I mean by extra magic, though. Specialization is "nothing new," but it's still relying on some ugly type machinery to accomplish it. I think specializing for `Handle = AllocHandle` is a sound form of specialization; HOWEVER, it's still relying on an incomplete feature, AND it's not actually quite the specialization you actually want; the actual specialization should be on `S: Allocator`, and that's unambiguously an unsound specialization (since impls could have lifetime bounds). So for the specialization to be sound, `Store` needs to have an additional precondition that `S::Handle = AllocHandle` implies `S: Allocator`.

But, having worked it, while it's still messy, I'm pleasantly surprised at it being a fairly straightforward application of encapsulated specialization.

<details><summary>roughly said impl magic</summary>

[[playground]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=897f2ee3cb1cf1d466293c157ae25901)

Using `ManuallyDrop<Box<T>>` as the unique handle repr because it currently has the desired semantics.

```rust
pub struct UniqueHandle<T: ?Sized, S: Store>(<S::Handle as UniqueHandleSpec<S>>::Output<T>);

impl<T: ?Sized, S: Store> UniqueHandle<T, S> {
    pub unsafe fn new(handle: S::Handle, meta: <T as Pointee>::Metadata) -> Self {
        UniqueHandle(<S::Handle as UniqueHandleSpec<S>>::Output::pack(
            handle, meta,
        ))
    }

    pub fn get(&self) -> S::Handle {
        <S::Handle as UniqueHandleSpec<S>>::Output::unpack(&self.0).0
    }

    pub fn meta(&self) -> <T as Pointee>::Metadata {
        <S::Handle as UniqueHandleSpec<S>>::Output::unpack(&self.0).1
    }
}

trait UniqueHandleSpec<S: Store> {
    type Output<T: ?Sized>: UniqueHandleImpl<T, S>;
}

trait UniqueHandleImpl<T: ?Sized, S: Store> {
    unsafe fn pack(handle: S::Handle, meta: <T as Pointee>::Metadata) -> Self;
    fn unpack(handle: &Self) -> (S::Handle, <T as Pointee>::Metadata);
}

impl<S: Store> UniqueHandleSpec<S> for S::Handle {
    default type Output<T: ?Sized> = (S::Handle, <T as Pointee>::Metadata, PhantomData<T>);
}

impl<T: ?Sized, S: Store> UniqueHandleImpl<T, S>
    for (S::Handle, <T as Pointee>::Metadata, PhantomData<T>)
{
    unsafe fn pack(handle: S::Handle, meta: <T as Pointee>::Metadata) -> Self {
        (handle, meta, PhantomData)
    }
    fn unpack(handle: &Self) -> (S::Handle, <T as Pointee>::Metadata) {
        (handle.0, handle.1)
    }
}

impl<S: Store<Handle = AllocHandle>> UniqueHandleSpec<S> for AllocHandle {
    type Output<T: ?Sized> = ManuallyDrop<Box<T>>;
}

impl<T: ?Sized, S: Store<Handle = AllocHandle>> UniqueHandleImpl<T, S> for ManuallyDrop<Box<T>> {
    unsafe fn pack(handle: AllocHandle, meta: <T as Pointee>::Metadata) -> ManuallyDrop<Box<T>> {
        ManuallyDrop::new(Box::from_raw(ptr::from_raw_parts_mut(
            handle.get().cast(),
            meta,
        )))
    }
    fn unpack(handle: &ManuallyDrop<Box<T>>) -> (AllocHandle, <T as Pointee>::Metadata) {
        // SAFETY: we're pretending to be std
        let handle: &NonNull<T> = unsafe { transmute(handle) };
        let ptr = handle.as_ptr();
        (AllocHandle::new(ptr.cast()).unwrap(), ptr::metadata(ptr))
    }
}
```

-----
</details>